### PR TITLE
[Placement Group]Placement Group supports gcs failover(Part2)

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
If GCS is restarted during placement group scheduling, raylet may have resource leakage. 
When GCS restarts, it doesn't know which bundles it has scheduled in the previous lifecycle. 
In this case, GCS will send a list of bundle ids that are still needed. And Raylet will release other bundles. 
If the node is dead, there is no need to send the request of release unused bundles.
At present, the worker does not have a bundle ID, so we cannot filter out the workers used by unused bundles. We will solve it in next pr(https://github.com/ray-project/ray/pull/12036).

## Related issue number

https://github.com/ray-project/ray/issues/9772

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by raulchen — APPROVED** [review]

**Comment by rkooo567** [comment]
Please don’t merge it until I review it!

**Comment by ffbin** [comment]
> Please don’t merge it until I review it!

Ok, thanks :)

**Comment by rkooo567** [comment]
Btw, this seems to fail after we merged part 1. test_create_placement_group_after_gcs_server_restarts. Is it expected?

**Comment by ffbin** [comment]
> Btw, this seems to fail after we merged part 1. test_create_placement_group_after_gcs_server_restarts. Is it expected?

It is not expected, can you share the CI job link, thanks.

**Comment by rkooo567** [comment]
https://flakey-tests.ray.io/

You can see after the first PR, the tests became extremely flaky.

**Comment by ffbin** [comment]
> https://flakey-tests.ray.io/
> 
> You can see after the first PR, the tests became extremely flaky.

I remove the timeout of ray.get, thanks.

**Comment by rkooo567** [comment]
Hmm this seems to diverge a lot from the original implementation (raylet just releases all prepared resources when gcs failures are detected). Isn't the original implementation significantly simpler? Why did we change it to this way?

**Comment by ffbin** [comment]
> Hmm this seems to diverge a lot from the original implementation (raylet just releases all prepared resources when gcs failures are detected). Isn't the original implementation significantly simpler? Why did we change it to this way?

The current implementation method is similar to the release of unused workers by GCS actor manager, and the implementation is relatively simple. Raylet cannot detect GCS restart at present, because GCS client does not disclose GCS restart information. Moreover, raylet detection is not accurate (such as raylet and GCS network jitter).

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 1]
```diff
@@ -470,6 +470,7 @@ void GcsPlacementGroupManager::LoadInitialData(const EmptyCallback &done) {
   auto callback = [this,
                    done](const std::unordered_map<PlacementGroupID,
                                                   rpc::PlacementGroupTableData> &result) {
+    std::unordered_map<NodeID, std::vector<rpc::Bundle>> node_to_bundles;
     for (auto &item : result) {
       auto placement_group = std::make_shared<GcsPlacementGroup>(item.second);
       if (item.second.state() != rpc::PlacementGroupTableData::REMOVED) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:476]: Btw, why do we add removed placement groups to the entry? I think we currently delete placement groups when they are deleted.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 2]
```diff
@@ -204,6 +204,17 @@ void GcsPlacementGroupScheduler::ScheduleUnplacedBundles(
     std::shared_ptr<GcsPlacementGroup> placement_group,
     std::function<void(std::shared_ptr<GcsPlacementGroup>)> failure_callback,
     std::function<void(std::shared_ptr<GcsPlacementGroup>)> success_callback) {
+  // We need to ensure that the PrepareBundleResources won't be sent before the reply of
+  // ReleaseUnusedBundles is returned.
+  if (!nodes_of_releasing_unused_bundles_.empty()) {
+    RAY_LOG(INFO) << "Failed to schedule placement group " << placement_group->GetName()
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:210]: Good catch!

**Review by rkooo567 — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:476]: We filtered out the removed placement groups.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:476]: _Oops..! lol. Sorry I was confused...

**Comment by rkooo567** [comment]
Can you make sure the skipped tests are passed in this PR or make sure you understand why the failures happen?

**Comment by ffbin** [comment]
> Can you make sure the skipped tests are passed in this PR or make sure you understand why the failures happen?

I tested it on the Mac and find that the failures happen is due to lightweight heartbeat.

**Pull request merged**